GSoC_19/posts/websocket-passive-scan-using-scripts-with-zap.org

```diff
@@ -0,0 +1,851 @@
+              var contexts = Model.getSingleton().getSession().getContexts();
+              var context;
+
+              for(var i = 0; i < contexts.size(); i++){
+                  context = contexts.get(i);
+                  var contextUsers = extUserManagment.getContextUserAuthManager(context.getIndex()).getUsers();
+                  if(contextUsers.size() > 0){
+                      for(var j = 0; j < contextUsers.size(); j++ ){
+                          usersList.push(contextUsers.get(j));
+                      }
+                  }
+              }
+          }
+          return usersList;
+      }
+
+      function getExtensionUserManagment(){
+          if(extUserManagment == null){
+              extUserManagment = Control.getSingleton()
+                  .getExtensionLoader()
+                  .getExtension(ExtensionUserManagement.class);
+          }
+          return extUserManagment;
+      }
+    #+END_SRC
+    
+    In addition to that with that we are using the =Appache Commons Codec=
+    library, which came bundled with ZAP, to hash the usernames.
+    
+    #+BEGIN_SRC js    
+var DigestUtils = Java.type("org.apache.commons.codec.digest.DigestUtils");
+    
+function getHashes(username){
+    var usernameHashes = {};
+    usernameHashes['MD2'] = new RegExp(DigestUtils.md2Hex(username), 'gmi');
+    usernameHashes['MD5'] = new RegExp(DigestUtils.md5Hex(username), 'gmi');
+    usernameHashes['SHA1'] = new RegExp(DigestUtils.sha1Hex(username), 'gmi');
+    usernameHashes['SHA256'] = new RegExp(DigestUtils.sha256Hex(username), 'gmi');
+    usernameHashes['SHA384'] = new RegExp(DigestUtils.sha384Hex(username), 'gmi');
+    usernameHashes['SHA512'] = new RegExp(DigestUtils.sha512Hex(username), 'gmi');
+
+    return usernameHashes;
+}
+    #+END_SRC 
+    
+    It's quite obvious that in the next step we should just check the above
+    patterns. This might seem difficult if not familiar with the code base,
+    however it's not (I am not familiar with code base too). The great about ZAP
+    is the fact that have plenty of passive and active rules available. 
+
+    Firstly, I would like to refer to community scripts. Community scripts are
+    developed by community and they are responsible for various "extra" rules
+    that ZAP has. If you don't use them already, I totally recommend to give them a try by
+    install them via =Marketplace=. In the Community Scripts you can find all
+    you need to develop a "sophisticated" WebSocket Passive Scanner script. In
+    fact, some of the above scripts are based on existing script (e.g. [[https://github.com/zaproxy/community-scripts/blob/master/passive/Find%2520Emails.js][Email
+    Disclosure]]).  
+
+    If you don't find what you are searching for in the Community Scripts, then maybe you
+    have to take look at the Java rules. Most of the above scripts are based on them. For
+    example, script in this use case is inspired and uses code blocks from the
```

Comment:
`script in this use case is inspired and uses code blocks from the` 
&gt; 
`the script in that last use case is inspired by and uses code blocks from the`